### DISCUSSION TITLE
grpc metadata based routing

### DISCUSSION THREAD

**Discussion by Syed-Shahidh-Ilhan** [argoproj/argo-rollouts#3902] [Q&A] [answered]
I'm working on implementing a canary deployment strategy for a gRPC service using Argo Rollouts and Istio. My goal is to route specific gRPC requests to the canary subset based on gRPC metadata.

Specifically, I want to achieve the following:

Route gRPC requests with a specific metadata (e.g., 'x-canary: true') to the canary subset.
All other gRPC requests should go to the stable subset.
After the rollout is complete, all traffic should be directed to the new stable version, regardless of the metadata.
Can someone provide guidance on how to configure Argo Rollouts to achieve this gRPC metadata-based routing? Is this possible using the built-in features of Argo Rollouts, or would I need to implement any custom logic?

**Comment by kostis-codefresh** [comment]
You might find these docs interesting

- https://rollouts-plugin-trafficrouter-gatewayapi.readthedocs.io/en/latest/features/grpc/
- https://rollouts-plugin-trafficrouter-gatewayapi.readthedocs.io/en/latest/features/header-based-routing/

**Comment by Syed-Shahidh-Ilhan** [comment]
Thank you for the suggestion, @kostis-codefresh. However, our current infrastructure doesn't incorporate the Gateway API. Instead, we utilize Argo Rollouts in combination with Istio as our service mesh. Within our Istio setup, we rely on Istio's native APIs, such as Gateway and VirtualService.

I'm curious if it's possible to achieve a similar solution while maintaining our existing architecture, without the need to migrate to the Gateway API. Is there a way to accomplish this using our current Istio and Argo Rollouts configuration?"

**Comment by kostis-codefresh** [comment, marked as answer]
Header based routing is also supported with plain Istio https://argo-rollouts.readthedocs.io/en/stable/features/traffic-management/#traffic-router-support-istio_1 

Regarding Grpc support I think you need to check the source code of Argo Rollouts....

**Comment by Syed-Shahidh-Ilhan** [comment]
Thanks a lot @kostis-codefresh , will try things out
